KAFKA-6016: Make the reassign partitions system test use the idempotent producer

apurvam opened the pull request:
With these changes, we are ensuring that the partitions being reassigned are from non-zero offsets. We also ensure that every message in the log has producerId and sequence number. 

This means that it successfully reproduces https://issues.apache.org/jira/browse/KAFKA-6003, as can be seen below:

```

[2017-10-05 20:57:00,466] ERROR [ReplicaFetcher replicaId=1, leaderId=4, fetcherId=0] Error due to (kafka.server.ReplicaFetcherThread)
kafka.common.KafkaException: Error processing data for partition test_topic-16 offset 682
        at kafka.server.AbstractFetcherThread$$anonfun$processFetchRequest$2$$anonfun$apply$mcV$sp$1$$anonfun$apply$2.apply(AbstractFetcherThread.scala:203)
        at kafka.server.AbstractFetcherThread$$anonfun$processFetchRequest$2$$anonfun$apply$mcV$sp$1$$anonfun$apply$2.apply(AbstractFetcherThread.scala:171)
        at scala.Option.foreach(Option.scala:257)
        at kafka.server.AbstractFetcherThread$$anonfun$processFetchRequest$2$$anonfun$apply$mcV$sp$1.apply(AbstractFetcherThread.scala:171)
        at kafka.server.AbstractFetcherThread$$anonfun$processFetchRequest$2$$anonfun$apply$mcV$sp$1.apply(AbstractFetcherThread.scala:168)
        at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
        at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
        at kafka.server.AbstractFetcherThread$$anonfun$processFetchRequest$2.apply$mcV$sp(AbstractFetcherThread.scala:168)
        at kafka.server.AbstractFetcherThread$$anonfun$processFetchRequest$2.apply(AbstractFetcherThread.scala:168)
        at kafka.server.AbstractFetcherThread$$anonfun$processFetchRequest$2.apply(AbstractFetcherThread.scala:168)
        at kafka.utils.CoreUtils$.inLock(CoreUtils.scala:218)
        at kafka.server.AbstractFetcherThread.processFetchRequest(AbstractFetcherThread.scala:166)
        at kafka.server.AbstractFetcherThread.doWork(AbstractFetcherThread.scala:109)
        at kafka.utils.ShutdownableThread.run(ShutdownableThread.scala:64)
Caused by: org.apache.kafka.common.errors.UnknownProducerIdException: Found no record of producerId=1000 on the broker. It is possible that the last message with the producerId=1000 has been removed due to hitting the retention limit.
```

Commit:
Initial commit trying to create the scenario where we are creating a
replica from scratch but starting from a non zero sequence when doing
so.

Commit:
Use retention bytes to force segment deletion

Commit:
Configure the log so that we can reproduce the case where we are building producer state from a non zero sequence

Commit:
Removed forced override of log.retention.check.interval.ms since it adds no value (it's the same as the Kafka default) and prevents overriding in tests

apurvam commented:
I can confirm that the fix in #4004 works: the exception above no longer occurs.

However, the test in the current from is flaky. Since segments are being deleted fairly fast, the verification fails since some messages are removed after acknowledgement but before the consumer consumes them, causing reports of data loss.

apurvam commented:
I am inclined to close this for now, because the approach of leveraging the reassign partitions test to validate that we can rebuild replicas from the middle of the log seems untenable. We did get a good data point out of this: that the fix in #4004 is effective in fixing the KAFKA-6003.

But building a reliable and useful test to exercise these code paths is trickier than originally thought. We would like to test we can build replicas from the middle of the log without losing data. 

But the only way to build replicas from something other than the first segment is to actually remove segments, which means we lose data. 

So there is doesn't seem to be a reliable way to reconcile these two requirements in the current approach. Will have to think of an alternate approach.

What do you think? @hachikuji @ijuma

ijuma commented:
@apurvam you could set the retention, wait for some segments to be deleted, increase the retention (it's a topic config), produce more data and then do the rest of the test. Would this work?

hachikuji commented:
I suggested also first writing some messages with old timestamps. These messages will be deleted and then we can do our validation on a set of messages with newer timestamps.

ijuma commented:
That's true

apurvam commented:
Hmm.. at first I thought the old timestamp approach would be better off in a separate test. But thinking more, I think the following can be done within the context of the reassign partitions test, and yet test reassign partitions properly:

1. We seed the topic with some old messages which clearly violate the retention.ms (messages which are 10 days old, say).
2. We wait until we are sure that one cleaning cycle has occurred. For instance, if we set the `log.retention.cleanup.interval.ms` to 10 seconds, we can wait around 15 seconds in the test.
3. After the wait, we front of the log should have been removed. Any reassignment started now should start from non zero offsets and non zero sequences.
4. We proceed with the current reassign partitions logic which does a produce consume validate for the duration of the reassignment. There should be no data loss in this case. 

The two changes we need for this are:

1. Update to the verifiable producer to specify the create time for each message.
2. Add a step to the reassign partitions test to ensure that the log starts from a non-zero offset before reassignment.

Step 2 could be an additional test case, which means that the changes are purely additive. 

The above scheme should work, and it _should_ be reliable as well. Thanks for the suggestion @hachikuji !

Commit:
Update the patch so that we seed the topic with older messages which are
going to be removed before reassignment starts. This way, the
reassignment starts from a non-zero offset, but we also have a valid
produce/consume/validate cycle through the course of the reassignment.

Further, the changes are purely additive, so the existing reassignment
tests execute in addition to a new set which tests reassignment starting
at a non-zero offset.

apurvam commented:
I updated the PR with the proposal above. Local runs show that it is stable, and also reproduces the original issue when I back out the fix. 

I started a a test run on jenkins: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/1096/

Commit:
Check for duplicates during verification

apurvam commented:
The jenkins run is looking good so far, but I am once more in doubt about why this works. Particularly, the way the test is setup, all of the first producer's messages should be deleted before reassignment starts, since we wait for one cleaning cycle after the last message is produced before starting the reassignment.

If the above is true, the original problem should not be reproduced, since we will copying only messages from the new producer, which start at sequence 0.

I guess another possibility is that because of the different cycles of cleanup on the replicas, the original replication process in the seed stage it self displays the error, because segments on the replica get cleaned up before the segments on the leader.

apurvam commented:
I looked over the failures when the patch was backed out. THe failure was indeed during the seeding stage, because a replica performed segment deletion before the leader. So the replication failed, and the producer got stuck because of a `NOT_ENOUGH_REPLICAS` exception.

apurvam commented:
Looks like the system test run is a success: http://confluent-kafka-branch-builder-system-test-results.s3-us-west-2.amazonaws.com/2017-10-06--001.1507355332--apurvam--KAFKA-6016-add-idempotent-producer-to-reassign-partitions--bf0b129/report.html

Also, this test reproduces the original case in KAFKA-6003 inadvertently. When we are seeding the topic with old messages, the replication of those message itself hits the `UnknownProducerId` error before the fix in KAFKA-6003. This is because of the asynchronous nature of segment removal where segments are removed on the leader and followers at differnt times. This leads to a gap in sequences on the follower during replication, causing the replication failure with the code before #4004 .

When the seed stage is finished and once the reassignment starts, all messages from the seed producer are removed. Retaining these messages would result in duplicates, which are no longer allowed by the test since it uses the idempotent producer. Thus the reassignment phase doesn't reproduce the original error. 

One additional benefit of this test is that it starts reassignment from non-zero offsets, which is probably a good code path to test in the long term.

cc @hachikuji @ijuma

apurvam commented:
retest this please

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/ProducerRecord.java:
```diff
@@ -58,7 +58,7 @@
      * @param topic The topic the record will be appended to
      * @param partition The partition to which the record should be sent
-     * @param timestamp The timestamp of the record
+     * @param timestamp The timestamp of the record, in milliseconds since epoch.
```
Since we're updating the javadoc, maybe we should mention the effect of using `null`?

apurvam replied:
What is the effect of null? That the server assigns the timestamp, which is the log append time?

hachikuji replied:
I think the effect is that the producer will assign the timestamp using `System.currentTime` before sending.

hachikuji commented on tests/kafkatest/services/kafka/templates/kafka.properties:
```diff
@@ -30,7 +30,6 @@ num.partitions=1
 num.recovery.threads.per.data.dir=1
 log.retention.hours=168
 log.segment.bytes=1073741824
-log.retention.check.interval.ms=300000
```
Why remove this? I'd have though that we'd make it configurable if anything.

apurvam replied:
Removing this makes it configurable (I override it in my test). Otherwise, these values are appended at the end of the properties file, which means they can't be changed in the tests.

hachikuji commented on tests/kafkatest/services/verifiable_producer.py:
```diff
@@ -125,8 +127,8 @@ def _worker(self, idx, node):
             self.logger.info("Setting up an idempotent producer")
-            producer_prop_file += "\nmax.in.flight.requests.per.connection=1\n"
-            producer_prop_file += "\nretries=50\n"
+            producer_prop_file += "\nmax.in.flight.requests.per.connection=2\n"
```
Why not just use the default of 5?

apurvam replied:
Will update this.

hachikuji commented on tools/src/main/java/org/apache/kafka/tools/VerifiableProducer.java:
```diff
@@ -144,6 +153,15 @@ private static ArgumentParser argParser() {
                 .metavar("CONFIG_FILE")
                 .help("Producer config properties file.");
 
+        parser.addArgument("--message-create-time")
```
I guess another way would be to let this configuration be a delta which is added to the the current time. That way we wouldn't have a lot of messages with the same timestamp, which might be a little uncommon in practice.

apurvam replied:
Messages are assigned create times _starting_ from this value. The time is incremented on every send by the interval between the sends. So they all won't have the same create time.

hachikuji replied:
Ah, ok. That works.

hachikuji commented on tests/kafkatest/tests/core/reassign_partitions_test.py:
```diff
@@ -38,13 +39,20 @@ def __init__(self, test_context):
+                                      "partitions": self.num_partitions,
+                                      "replication-factor": 3,
+                                      'configs': {
+                                          "min.insync.replicas": 3,
```
Was it intentional to increase this?

apurvam replied:
Yes. It is a better test. With min.isr = 2, the produce/consume/validate would work even if one replica was offline, which is not what we want.

hachikuji replied:
Maybe worth a comment explaining this?

hachikuji commented on tests/kafkatest/tests/core/reassign_partitions_test.py:
```diff
@@ -86,28 +94,60 @@ def reassign_partitions(self, bounce_brokers):
+        # Since the configured check interval is 5 seconds, we wait another
+        # 6 seconds to ensure that at least one more cleaning so that the last
+        # segment is deleted.
+        time.sleep(6)
```
Alternatively, we could poll the start offset and wait for it to match the end offset.

apurvam replied:
Yes. That was my plan B, if the test in its current form was flakey. But that would be a bit of extra work since this information is not easily exposed by the current console consumer or verifiable consumer.

But since we got 120 successful runs of this test, the current timeouts are sufficient IMO.

hachikuji replied:
Fair enough. A bit annoying sometimes that we don't have the full APIs to work with.

apurvam replied:
I added a comment to the code to mention the alternate, more robust, way to do this. We can execute on it if/when the test starts flaking out.

hachikuji reviewed (commented):
Thanks for the extra testing. Definitely good to exercise these paths. Left a few comments/suggestions.

Commit:
Address PR comments

Commit:
Update test comments per PR recommendations

Commit:
Fix checkstyle failure

hachikuji reviewed (approved):
LGTM
